Lib/pdb.py

```diff
@@ -96,17 +96,47 @@ class Restart(Exception):
 __all__ = ["run", "pm", "Pdb", "runeval", "runctx", "runcall", "set_trace",
            "post_mortem", "help"]
 
+
+def find_first_executable_line(code):
+    """ Try to find the first executable line of the code object.
+
+    Equivalently, find the line number of the instruction that's
+    after RESUME
+
+    Return code.co_firstlineno if no executable line is found.
+    """
+    prev = None
+    for instr in dis.get_instructions(code):
+        if prev is not None and prev.opname == 'RESUME':
+            if instr.positions.lineno is not None:
+                return instr.positions.lineno
+            return code.co_firstlineno
+        prev = instr
+    return code.co_firstlineno
+
 def find_function(funcname, filename):
     cre = re.compile(r'def\s+%s\s*[(]' % re.escape(funcname))
     try:
         fp = tokenize.open(filename)
     except OSError:
         return None
+    funcdef = ""
+    funcstart = None
     # consumer of this info expects the first line to be 1
     with fp:
         for lineno, line in enumerate(fp, start=1):
             if cre.match(line):
-                return funcname, filename, lineno
+                funcstart, funcdef = lineno, line
+            elif funcdef:
+                funcdef += line
+
+            if funcdef:
+                try:
+                    funccode = compile(funcdef, filename, 'exec').co_consts[0]
```

Comment:
You could, instead of calling compile many times, build funcdef up to the first line dedented to the "def" (or EOF, is that comes first).

Reply:
Will that be faster in general? In most cases, the current solution will resolve on the first statement of the function(it does not require the following lines). The function body could be long - and we are iterating line by line from a *file*. So there's IO cost, Python-level loop cost, (we can save the string concatination cost with a list but that's also something). Meanwhile, `compile` is a C-level function and it might be faster.

I don't have a very precise estimation on the performance of `compile` function on small piece of code, is my thought way off?